devops/scripts/benchmarks/utils/flamegraph.py

```diff
@@ -0,0 +1,436 @@
+            if not os.access(script_path, os.X_OK):
+                raise RuntimeError(f"FlameGraph script not executable: {script_path}")
+            log.debug(f"Verified {script} exists and is executable")
+
+        # Store repo_dir for later use when generating flamegraphs
+        self.repo_dir = repo_dir
+
+        log.info("FlameGraph tools ready.")
+
+        if options.results_directory_override == None:
+            self.flamegraphs_dir = os.path.join(
+                options.workdir, "results", "flamegraphs"
+            )
+        else:
+            self.flamegraphs_dir = os.path.join(
+                options.results_directory_override, "flamegraphs"
+            )
+
+    def _prune_flamegraph_dirs(self, res_dir: str, FILECNT: int = 10):
+        """Keep only the last FILECNT files in the flamegraphs directory."""
+        files = os.listdir(res_dir)
+        files.sort()  # Lexicographical sort matches timestamp order
+        if len(files) > 2 * FILECNT:
+            for f in files[: len(files) - 2 * FILECNT]:
+                full_path = os.path.join(res_dir, f)
+                if os.path.isdir(full_path):
+                    shutil.rmtree(full_path)
+                else:
+                    os.remove(full_path)
+                    log.debug(f"Removing old flamegraph file: {full_path}")
+
+    def cleanup(self, bench_cwd: str, perf_data_file: str):
+        """
+        Remove incomplete output files in case of failure.
+        """
+        flamegraph_dir = os.path.dirname(perf_data_file)
+        flamegraph_base = os.path.basename(perf_data_file)
+        for f in os.listdir(flamegraph_dir):
+            if f.startswith(flamegraph_base + "."):
+                os.remove(os.path.join(flamegraph_dir, f))
+                log.debug(f"Cleanup: Removed {f} from {flamegraph_dir}")
+
+    def setup(
+        self, bench_name: str, command: list[str], extra_perf_opt: list[str] = None
+    ):
+        """
+        Prepare perf data file name and full command for the benchmark run.
+        Returns a tuple of (perf_data_file, perf_command).
+        """
+        # Check if perf is available
+        if not shutil.which("perf"):
+            raise FileNotFoundError(
+                "perf command not found. Please install linux-tools or perf package."
+            )
+
+        os.makedirs(self.flamegraphs_dir, exist_ok=True)
+        bench_dir = os.path.join(f"{self.flamegraphs_dir}", f"{bench_name}")
+
+        os.makedirs(bench_dir, exist_ok=True)
+
+        perf_data_file = os.path.join(
+            bench_dir, f"{self.timestamp}_{options.save_name}.perf.data"
+        )
+
+        if extra_perf_opt is None:
+            extra_perf_opt = []
+
+        # Default perf record options for flamegraph generation
+        perf_command = (
+            [
+                "perf",
+                "record",
+                "-g",  # Enable call-graph recording
+                "-F",
+                "99",  # Sample frequency
```

Comment:
this seems low. we should experiment with different values and pick what gives us the best flamecharts.